### DISCUSSION TITLE
ascii timeline: mix states and words in one timeline

### DISCUSSION THREAD

**Discussion by Mark2Mark** [coldtype/coldtype#127] [General]
Hi Rob,

I have puzzled together a little piece that tries to mix font style states and words in one timeline. It feels quite fragile and partly I don’t even know what I was doing exactly.

1) I wonder if there is a better way to achieve that same result.

```py
from coldtype import *

fnt = Font.MutatorSans()

states = {
	"A": dict(wdth=1, wght=0),
	"B": dict(wdth=1, wght=1),
	"C": dict(wdth=0, wght=1),
	"D": dict(wdth=0, wght=0),
}

at = AsciiTimeline(4, 90, """
                                            <
 [A       ][B       ][C       ][D           ]
YYY_CCC 
          *TT_DD
                    *LLLLL
                             *O_O_O
""",
states).shift("end", -8)

def styler(c):
	if c.text.startswith(" "):
		return "", Style(fnt, 140)
	else:
		return c.text.replace("_", " "), Style(fnt, 140, wght=0)

@animation((1080, 1080), timeline=at, bg=0)
def specimen_01(f):

	state = at.kf("eeio", keyframes=states)

	return (
		f.t.words.currentGroup().pens(f, styler)
			.f(1)
			.align(f.a.r.inset(870), "cx", "mxy")
			,
		Glyphwise("COLDTYPE", lambda g: [
		Style(fnt, 200, **{**state})
		, at.kf("eeio", f.i)
		])
		.f(1)
		.align(f.a.r)
	)
```

2) What is this funky `**{**state}` thing doing? I had it from an example file, where it was sth like `**{**state, **spacing}`

3) My if–else condition in the style method seems hacky. Is there a better way? Also for getting more than one words to show, instead of doing my replace-undescore-hack?

4) The ascii timeline can take some modifier characters, I have seen `≈` , `<` ,`>` , `|` ,`*` and the bullet, but from the examples as well as the ascii timeline source code in the lib, i can’t figure out what they exactly do. Here I seem to need to have to add the asterisk in order for the words to change and not keep showing the first one all the time. Also the words need to be placed before the style state changes to show. That is why I needed to add one space in front of the `[A   ]`

**Comment by stenson** [comment]
Great question! Using the `.words` functionality on the timeline is kind of an all-or-nothing affair, since there is no concept of "tracks" in the AsciiTimeline (so you can better use multiple lines to accomodate ascii text editing).

So, I'd say there are two approaches to what you're trying to accomplish:

1. Use multiple AsciiTimelines
2. Use the .style syntax that the .words functionality supports (as seen in the `examples/animations/ascii_words.py` example)

Here's approach 1, with two distinct timelines. The main "gotcha" here is that you'll need to manually `.hold` whichever timeline isn't being passed to the decorator (when you pass a `timeline=` to the decorator, it automatically `.hold`s for you).

```python
from coldtype import *

fnt = Font.MutatorSans()

states = {
	"A": dict(wdth=1, wght=0),
	"B": dict(wdth=1, wght=1),
	"C": dict(wdth=0, wght=1),
	"D": dict(wdth=0, wght=0),
}

track1 = """
                                            <
 [A       ][B       ][C       ][D           ]
"""
track2 = """
                                            <
*YYY CCC  •
          *TT DD    •
                    *LLLLL   •
                             *O O O        •
"""

at = AsciiTimeline(4, 60, track1).shift("end", -8)
at2 = AsciiTimeline(4, 60, track2)

def styler(c):
	return c.text, Style(fnt, 140, wght=0)

@animation((1080, 1080), timeline=at, bg=0)
def specimen_01(f):
    at.hold(f.i)
    at2.hold(f.i)

    state = at.kf("eeio", keyframes=states)

    return P(
        at2.words.currentGroup().pens(f, styler)
            .f(1)
            .align(f.a.r.inset(870), "cx", "mxy")
            .removeFutures()
            ,
        StSt("COLDTYPE", Style(fnt, 100, **state))
            .f(1)
            .align(f.a.r))
```

Upside to that approach: more flexible. Downside: having to make sure the two timelines are in sync.

And here's approach 2:

```python
from coldtype import *

fnt = Font.MutatorSans()

states = {
	".A": dict(wdth=1, wght=0),
	".B": dict(wdth=1, wght=1),
	".C": dict(wdth=0, wght=1),
	".D": dict(wdth=0, wght=0),
}

track2 = """
                                            <
 [.A      ][.B      ][.C      ][.D         ]
*YYY CCC  •
          *TT DD    •
                    *LLLLL   •
                             *O O O        •
"""

at = AsciiTimeline(4, 60, track2)

for t in at.timeables:
    if t.name.startswith("."):
        t.end = t.end - 8

def styler(c):
	return c.text, Style(fnt, 140, wght=0)

@animation((1080, 1080), timeline=at, bg=0)
def specimen_01(f):
    state = at.kf("eeio", keyframes=states)

    return P(
        at.words.currentGroup().pens(f, styler)
            .f(1)
            .align(f.a.r.inset(870), "cx", "mxy")
            .removeFutures()
            ,
        StSt("COLDTYPE", Style(fnt, 100, **state))
            .f(1)
            .align(f.a.r)
        )
```

Upside to that approach: it's just one timeline. Downside: a little clunkier to do the "end" manipulation in a for-loop. All-in-all though, I much prefer the second approach. The .style syntax exists to allow you to modify words with timeline metadata, but here we're just ignoring those styles as far as the .words text-setting goes.

As for the special symbols, for any kind of the `.words` functionality, when these symbols begin a bit of text:

- `*` means begin a new sequence of text (i.e. don't join to text before this on the timeline)
- `•` (option+8) means end a sequence of text (i.e. clear-the-screen of any text)
- `≈` means inset a newline before the characters that follow
- `+` means join the following characters to the previous ones without a space
- And no-special-character before characters means join them to the previous ones with a space (i.e. the default case, words in a sentence)


And to answer the other questions:

2. `**` is python means use-dict-as-kwargs, which is a little obscure, but in this case, Style accepts a kwargs-dict of styling properties, so if we want to pass a dict to the Style constructor, you have to `**` it first. (I think of it like pouring out a box of legos on a table or something, i.e. taking everything out of the box and passing it one-by-one. The same thing works for passing a dict to another dict, i.e. to combine dicts. So the outer `**` is kwargs-ing the dictionary, and the inner `**` is kwargs-ing the dictionary being passed to the dict literal braces. In this case, the inner `**` are unnecessary. You're more likely to see that idiom when combining two dictionaries, e.g.:

```python
a = {"hello": "world"}
b = {1: 2}
c = {**a, **b}
```

`c` is now: `{'hello': 'world', 1: 2}`

3. Taking out spaces should be unnecessary if you're using the .words functionality as intended, as you can see in the modified versions. (I realize it's pretty hard to use as intended though, since it's completely undocumented!)

(Also, just out of curiosity, did you mean to specify 90 fps as the frame rate?)

**Reply by Mark2Mark** [reply]
Thanks Rob! I’ll look into those and come back with the results. Happy to give those a try!

Quickly for the 90 fps: no, I didn’t intend **_that_** actually. My words were too long to fit in the ascii timeline, so I increased the timeline segments and hence the speed.
```
           <
[A   ][B   ]
WordOne    WordTwo
```
That was probably a bit silly of me. Maybe there’s a better way?

**Reply by Mark2Mark** [reply]
Actually I found a better way to fit long text into a short/small ascii timeline by using a dictionary for the texts, that way the keys can be short and fit into the timeline well:
```py
at = AsciiTimeline(4, 30, """
                                <
 [A     ][B     ][C     ][D     ]
t1      •
        *t2     •
                *t3     •
                        *t4
""",
states).shift("end", -8)

textDict = {
    "t1" : "Text 1 here",
    "t2" : "Another bit",
    "t3" : "More and more words",
    "t4" : "Finally this piece of stuff.",
}

def styler(c):
    if c.text.startswith(" "):
        return "", Style(fnt, 40)
    else:
        return textDict[c.text], Style(fnt, 40)
```

The `[A   ]` bit still needs to be one space behind the `t1` key though, not terribly bad, but feels like that shouldn’t be necessary?

**Reply by stenson** [reply]
Oh good idea on the `textDict`! Actually in that case, if you don't need to worry about animating word-by-word, the `.words` functionality is pretty overengineered. You could avoid all of the special syntax (*•, etc.) and just use the standard bracketing like `[t1   ]` etc. Then you wouldn't have an issue with the leading space on the other ones either.

So the code complexity could be reduced quite a bit (no need for a `styler` or anything like that).

```python
from coldtype import *

at = AsciiTimeline(4, 30, """
                                <
[A   ]  [B   ]  [C   ]  [D   ]
[t1    ][t2    ][t3    ][t4    ]
""")

textDict = {
    "t1" : "Text 1 here",
    "t2" : "Another bit",
    "t3" : "More and more words",
    "t4" : "Finally this piece of stuff.",
}

states = {
    "A": dict(wdth=1, wght=0),
    "B": dict(wdth=1, wght=1),
    "C": dict(wdth=0, wght=1),
    "D": dict(wdth=0, wght=0),
}

@animation((1080, 1080), timeline=at, bg=0)
def specimen_01(f):
    state = at.kf("eeio", keyframes=states)
    textKey = at.current(track=1).t[0].name

    return P(
        StSt(textDict[textKey], Font.RecMono(), 50)
            .f(1)
            .align(f.a.r.inset(870), "cx", "mxy"),
        StSt("COLDTYPE", Font.MuSan(), 100, **state)
            .f(1)
            .align(f.a.r))
```
